### PR TITLE
add index key and size map to segment metadata

### PR DESCRIPTION
**Pull request by walterddr**
Description
===
related to #8862. we also want to report these info when hitting server segment metadata endpoint.

Details
===
This endpoint adds Index metadata to the `ColumnMetadata` (only size info)

This info will be included on either 
- `curl -X GET "https://localhost:9000/segments/{tableName}/{segmentName}/megadata?columns=*`
- `curl -X GET "https://localhost:9000/segments/{tableName}/metadata?columns=*`

Result looks like:
Screen Shot 2022-11-02 at 4 10 36 PM

Todo / Follow ups
===
- showing one column info per-segment might not be so useful. we can also return aggregated index info in `/segments/{tableName}/metadata` endpoint (when segmentName is not specified)
- `startree` index file is not included, will include in next PR

### CODE REVIEW

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/ColumnMetadataImpl.java [hunk 1]
```diff
@@ -65,13 +66,14 @@ public class ColumnMetadataImpl implements ColumnMetadata {
   private final int _totalNumberOfEntries;
   private final PartitionFunction _partitionFunction;
   private final Set<Integer> _partitions;
+  private final Map<ColumnIndexType, Long> _indexSizeMap;
```
Comment (walterddr) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/ColumnMetadataImpl.java:69]: adding this map here should allow us to report index info (only size for now)

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/SegmentMetadataImpl.java [hunk 2]
```diff
@@ -209,7 +216,8 @@ private void loadCreationMeta(InputStream crcFileInputStream)
     }
   }
 
-  private void init(PropertiesConfiguration segmentMetadataPropertiesConfiguration) {
+  private void init(PropertiesConfiguration segmentMetadataPropertiesConfiguration,
+      PropertiesConfiguration segmentIndexPropertiesConfiguration) {
```
Comment (walterddr) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/SegmentMetadataImpl.java:220]: no implemented yet: still need to parse this index PropertiesConfiguration but it should have everything

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/ColumnMetadataImpl.java (hunk 1 above)
Comment (walterddr) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/ColumnMetadataImpl.java:69]: planning to move IndexKey or IndexEntry (without the buffer) from segment-local to segment-spi. thoughts?

Comment (walterddr) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/ColumnMetadataImpl.java:69]: resolved. no need for this

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/SegmentMetadataImpl.java [hunk 3]
```diff
@@ -245,6 +248,26 @@ private void init(PropertiesConfiguration segmentMetadataPropertiesConfiguration
...
+    if (_segmentVersion == SegmentVersion.v3) {
+      File indexMapFile = new File(_indexDir, "v3" + File.separator + V1Constants.INDEX_MAP_FILE_NAME);
+      if (indexMapFile.exists()) {
+        PropertiesConfiguration mapConfig = CommonsConfigurationUtils.fromFile(indexMapFile);
+        for (String key : CommonsConfigurationUtils.getKeys(mapConfig)) {
+          try {
+            String[] parsedKeys = ColumnIndexUtils.parseIndexMapKeys(key, _indexDir.getPath());
+            if (parsedKeys[2].equals(ColumnIndexUtils.MAP_KEY_NAME_SIZE)) {
+              ColumnIndexType columnIndexType = ColumnIndexType.getValue(parsedKeys[1]);
+              _columnMetadataMap.get(parsedKeys[0]).getIndexSizeMap().put(columnIndexType, mapConfig.getLong(key));
```
Comment (npawar) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/SegmentMetadataImpl.java:262]: do we need a nullcheck on the value of get() here? could it be that something is in index_map but not in metadata.properties (due to a column remove/index remove)?

**Review by npawar — APPROVED** [review]
can we add a test in TablesResourceTest getOfflineSegmentsMetadata?

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/SegmentMetadataImpl.java (hunk 3 above)
Comment (walterddr) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/SegmentMetadataImpl.java:262]: i believe it is already caught in the Exception below?

**Comment by walterddr** [comment]
CC @npawar @nizarhejazi

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/SegmentMetadataImpl.java (hunk 3 above)
Comment (npawar) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/SegmentMetadataImpl.java:262]: ah missed that 👍

**Pull request merged**